### DISCUSSION TITLE
Git cloning within an extension silently fails on MacOS (clones everything but the `.git` folder).

### DISCUSSION THREAD

**Discussion by zeluspudding** [microsoft/vscode-discussions#850] [Extension Development QnA] [answered]
**TLDR** Does MacOS prohibit VS Code extensions from correctly adding hidden files to the file system? If not, does anyone know why `simple-git` fails to correctly clone git repos on MacOS but not on Windows?

I have an extension I've launched that works well on Windows machines. However, on MacOS, I recently discovered that git cloning fails in the same way using two different methods. Whether using `simple-git`:
```
const simpleGit = require('simple-get')
async function initialize_git(local_repo_path, remote_url= '') {
    const options = {
        baseDir: local_repo_path,
        binary: 'git',
        maxConcurrentProcesses: 6,
        trimmed: false,
    }
    git = simpleGit(options)
    const isRepo = await git.checkIsRepo()
    if (isRepo == false) {
        await git.clone(remote_url, local_repo_path)
    }
    return git
}
```

or using `spawnSync`
```
const {spawnSync}= require('child_process')
async function initialize_git(local_repo_path, remote_url= '') {
    const options = {
        baseDir: local_repo_path,
        binary: 'git',
        maxConcurrentProcesses: 6,
        trimmed: false,
    }
    git = simpleGit(options)
    const isRepo = await git.checkIsRepo()
    if (isRepo == false) {
        spawnSync('git', ['clone', remote_url, local_repo_path])
    }
    return git
}
```

the repository that's cloned will contain the remote's source files as expected but will not contain the actual git files (those within the hidden `.git` folder). As a result, the cloned repository is unusable for version control (because it's as if the folder was never `git init`ed). This failure is silent so my extension continues to execute until trying some git based action on the local repo and fails because said repo isn't technically a **git** repo.

As a workaround, I tried to see if I could use VS Code's internal `git.clone` command, like:
```
const vscode= require('vscode')
async function initialize_git(local_repo_path, remote_url= '') {
    const options = {
        baseDir: local_repo_path,
        binary: 'git',
        maxConcurrentProcesses: 6,
        trimmed: false,
    }
    git = simpleGit(options)
    const isRepo = await git.checkIsRepo()
    if (isRepo == false) {
        await vscode.commands.executeCommand('git.clone', remote_url, local_repo_path), 
    }
    return git
}
```

And this correctly clones the git repo (including the `.git` folder) but, sadly, I can't use this command because it results in showing a prompt to my end users (which I cannot seem to hide) that confuses our UX.  ...This got me thinking maybe the underlying issue with the `simple-git` and `spawnSync` approaches has to do with some MacOS file permission restrictions prohibiting VS Code extensions from adding hidden files to the file system. **Does MacOS prohibit VS Code extensions from correctly adding hidden files to the file system? If so, what's the work around? If not, does anyone know why `simple-git` fails to correctly clone git repos on MacOS but not on Windows?**

**Comment by r3m0t** [comment]
Check the Output Window for the Git extension, it will show the clone command(s) being used with all their flags. Maybe one of the flags is relevant.

**Comment by zeluspudding** [comment, marked as answer]
That was a good recommendation, as I hadn't thought to do so myself. It helped me track down the real issue which was that - due to how file paths work on MacOS (versus on Windows), my extension was accidentally deleting the `.git` folder. Ultimately git clone works as expected, but other parts of my application were failing. Thanks again for chiming in and providing that direction!
